File: articles.jdbc/src/test/java/pet/db/jdbc/tool/generator/RegistrationPayloadTestDataGenerator.java
package pet.db.jdbc.tool.generator;

import lombok.RequiredArgsConstructor;

import org.springframework.stereotype.Component;

import pet.db.jdbc.model.dto.payload.RegistrationPayload;
import pet.db.jdbc.model.dto.User;

import java.util.List;

@Component
@RequiredArgsConstructor
public class RegistrationPayloadTestDataGenerator implements TestDataGenerator<RegistrationPayload> {

    private final TestDataGenerator<User> userTestDataGenerator;

    @Override
    public RegistrationPayload generateSavedData() {
        return generateSavedData(1).getFirst();
    }

    @Override
    public RegistrationPayload generateUnsavedData() {
        return generateUnsavedData(1).getFirst();
    }

    @Override
    public List<RegistrationPayload> generateSavedData(Integer dataSize) {
        return userTestDataGenerator.generateSavedData(dataSize).stream()
                .map(this::convertToRegistrationPayload)
                .toList();
    }

    @Override
    public List<RegistrationPayload> generateUnsavedData(Integer dataSize) {
        return userTestDataGenerator.generateUnsavedData(dataSize).stream()
                .map(this::convertToRegistrationPayload)
                .toList();
    }

    private RegistrationPayload convertToRegistrationPayload(User user) {
        return new RegistrationPayload(
                user.getUsername(),
                user.getEmail(),
                user.getPassword());
    }

}

File: src/test/kotlin/pet/articles/test/tool/generator/RegistrationPayloadTestDataGenerator.kt
package pet.articles.test.tool.generator

import org.koin.core.annotation.Named
import org.koin.core.annotation.Single

import pet.articles.model.dto.User
import pet.articles.model.dto.payload.RegistrationPayload

class RegistrationPayloadTestDataGenerator(
    private val userTestDataGenerator: TestDataGenerator<User>
) : TestDataGenerator<RegistrationPayload> {

    override fun generateSavedData(dataSize: Int): List<RegistrationPayload> =
        userTestDataGenerator.generateSavedData(dataSize).map(::convertToRegistrationPayload)

    override fun generateUnsavedData(dataSize: Int): List<RegistrationPayload> =
        userTestDataGenerator.generateUnsavedData(dataSize).map(::convertToRegistrationPayload)

    override fun generateInvalidData(): RegistrationPayload =
        convertToRegistrationPayload(userTestDataGenerator.generateInvalidData())

    private fun convertToRegistrationPayload(user: User): RegistrationPayload =
        RegistrationPayload(
            username = user.username,
            email = user.email,
            password = user.password!!
        )
}